File: reference.net/db4o-reference-chapters/src/com/db4odoc/clientserver/ExtClientExample.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using Db4objects.Db4o;
using Db4objects.Db4o.Ext;

namespace Db4objects.Db4odoc.Clientserver
{
	public class ExtClientExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public readonly static string ExtFileName ="formula1e.yap";
	
		public static void Main(string[] args) 
		{
			SwitchExtClients();
		}
		// end Main

		public static void SwitchExtClients() 
		{
			File.Delete(YapFileName);
			File.Delete(ExtFileName);
			IObjectServer server=Db4oFactory.OpenServer(YapFileName,0);
			try 
			{
				IObjectContainer client=server.OpenClient();
				Car car = new Car("BMW");
				client.Set(car);
				System.Console.WriteLine("Objects in the Main database file:");
				RetrieveAll(client);
			
				System.Console.WriteLine("Switching to additional database:");
				IExtClient clientExt = (IExtClient)client;
				clientExt.SwitchToFile(ExtFileName);
				car = new Car("Ferrari");
				clientExt.Set(car);
				RetrieveAll(clientExt);
				System.Console.WriteLine("Main database file again: ");
				clientExt.SwitchToMainFile();
				RetrieveAll(clientExt);
				clientExt.Close();
			}
			finally 
			{
				server.Close();
			}
		}
		// end SwitchExtClients

		public static void RetrieveAll(IObjectContainer db) 
		{
			IObjectSet result = db.Query(typeof(Car));
			ListResult(result);
		}
		// end RetrieveAll

		public static void ListResult(IObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}

File: reference.vb.net/Db4objects.Db4odoc/Clientserver/ExtClientExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports System.IO
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Ext


Namespace Db4objects.Db4odoc.ClientServer
    Public Class ExtClientExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"
        Public Shared ReadOnly ExtFileName As String = "formula1e.yap"

        Public Shared Sub Main(ByVal args() As String)
            SwitchExtClients()
        End Sub
        ' end Main

        Public Shared Sub SwitchExtClients()
            File.Delete(YapFileName)
            File.Delete(ExtFileName)
            Dim server As IObjectServer = Db4o.OpenServer(YapFileName, 0)
            Try
                Dim client As IObjectContainer = server.OpenClient()
                Dim car As Car = New Car("BMW")
                client.Set(car)
                System.Console.WriteLine("Objects in the Main database file:")
                RetrieveAll(client)

                System.Console.WriteLine("Switching to additional database:")
                Dim clientExt As IExtClient = CType(client, IExtClient)
                clientExt.SwitchToFile(ExtFileName)
                car = New Car("Ferrari")
                clientExt.Set(car)
                RetrieveAll(clientExt)
                System.Console.WriteLine("Main database file again: ")
                clientExt.SwitchToMainFile()
                RetrieveAll(clientExt)
                clientExt.Close()
            Finally
                server.Close()
            End Try
        End Sub
        ' end SwitchExtClients

        Public Shared Sub RetrieveAll(ByVal db As IObjectContainer)
            Dim result As IObjectSet = db.Get(GetType(Car))
            ListResult(result)
        End Sub
        ' end RetrieveAll

        Public Shared Sub DeleteAll(ByVal db As IObjectContainer)
            Dim result As IObjectSet = db.Get(GetType(Car))
            Dim item As Object
            For Each item In result
                db.Delete(item)
            Next
        End Sub
        ' end DeleteAll

        Public Shared Sub ListResult(ByVal result As IObjectSet)
            Console.WriteLine(result.Count)
            For Each item As Object In result
                Console.WriteLine(item)
            Next
        End Sub
        ' end ListResult
    End Class
End Namespace
